src/lc.rs

```diff
@@ -0,0 +1,433 @@
+use std::ops::{Add, Sub};
+
+use ff::Field;
+use pairing::Engine;
+
+use crate::multiexp::DensityTracker;
+
+/// Represents a variable in our constraint system.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Variable(pub(crate) Index);
+
+impl Variable {
+    /// This constructs a variable with an arbitrary index.
+    /// Circuit implementations are not recommended to use this.
+    pub fn new_unchecked(idx: Index) -> Variable {
+        Variable(idx)
+    }
+
+    /// This returns the index underlying the variable.
+    /// Circuit implementations are not recommended to use this.
+    pub fn get_unchecked(&self) -> Index {
+        self.0
+    }
+}
+
+/// Represents the index of either an input variable or
+/// auxiliary variable.
+#[derive(Copy, Clone, PartialEq, Debug, Eq, Hash)]
+pub enum Index {
+    Input(usize),
+    Aux(usize),
+}
+
+/// This represents a linear combination of some variables, with coefficients
+/// in the scalar field of a pairing-friendly elliptic curve group.
+#[derive(Clone)]
+pub struct LinearCombination<E: Engine> {
+    inputs: Indexer<E::Fr>,
+    aux: Indexer<E::Fr>,
+}
+
+#[derive(Clone)]
+struct Indexer<T> {
+    values: Vec<(usize, T)>,
+    // (index, key) of the last insertion operation
+    last_inserted: Option<(usize, usize)>,
+}
+
+impl<T> Default for Indexer<T> {
+    fn default() -> Self {
+        Indexer {
+            values: Vec::new(),
+            last_inserted: None,
+        }
+    }
+}
+
+impl<T> Indexer<T> {
+    pub fn from_value(index: usize, value: T) -> Self {
+        Indexer {
+            values: vec![(index, value)],
+            last_inserted: Some((0, index)),
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (&usize, &T)> + '_ {
+        self.values.iter().map(|(key, value)| (key, value))
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (&mut usize, &mut T)> + '_ {
+        self.values.iter_mut().map(|(key, value)| (key, value))
+    }
+
+    pub fn insert_or_update<F, G>(&mut self, key: usize, insert: F, update: G)
+    where
+        F: FnOnce() -> T,
+        G: FnOnce(&mut T),
+    {
+        if let Some((last_index, last_key)) = self.last_inserted {
```

Comment:
It took me a while to understand what's going on. So perhaps a comment would be good. If i understand it correctly the whole `if` part is an optimization for a case that often happens, which prevents doing a full binary search.

Reply:
done